src/Microsoft.ML.OnnxTransform/OnnxUtils.cs

```diff
@@ -98,239 +102,151 @@ public static OnnxModel CreateFromBytes(byte[] modelBytes)
-            }
-            else if (typeof(T) == typeof(System.Int64))
-            {
-                return new Tensor((System.Int64[])(object)data.ToArray(), shape.ToArray());
-            }
-            throw new NotImplementedException($"Not implemented type {typeof(T)}");
+            var typeMap = SystemTypeToOnnxType();
+            if (!typeMap.ContainsKey(typeof(T)))
+                throw new NotImplementedException($"Not implemented type {typeof(T)}");
+            return NamedOnnxValue.CreateFromTensor<T>(name, new DenseTensor<T>(data.ToArray(), shape.Select(x => (int)x).ToArray()));
         }
 
         /// <summary>
-        /// Sonoma API only provides CopyTo() functions with overloaded versions
-        /// based on data type. ML.NET cannot use the overloaded version and requires
-        /// generic version. CopyTo&lt;T&gt; is generic wrapper on top of
-        /// overloaded Tensor.CopyTo(List&lt;T&gt; dst) methods.
-        /// Also Tensor.CopyTo(List&lt;T&gt; dst) requires a list input, whereas ML.NET
-        /// provides array buffers to copy values to. This mismatch causes an extra copy.
+        /// Converts a Onnx type, that follows the System.Type convention
+        /// to the type system ML.NET recognizes (e.g. I4, I8, R4 etc.)
         /// </summary>
-        public static unsafe void CopyTo<T>(Tensor tensor, Span<T> dst)
+        /// <param name="type"></param>
+        /// <returns></returns>
+        public static PrimitiveType OnnxToMlNetType(System.Type type)
         {
-            var typeMap = SystemTypeToOnnxType();
-            if (typeMap.ContainsKey(typeof(T)))
-            {
-                if (tensor.GetDataType() != typeMap[typeof(T)])
-                {
-                    throw new InvalidOperationException( string.Format("Cannot copy source tensor of type {0} to managed type {1}.", tensor.GetDataType(), typeof(T)));
-                }
-                Span<T> tensorSpan = new Span<T>(tensor.UnsafeGetData().ToPointer(), tensor.GetSize());
-                tensorSpan.CopyTo(dst);
-                // TODO: the CopyTo() function is susceptible to GC reclaiming tensor
-                // during the method call. Use KeepAlive for now, and remove
-                // after permanent fix in CopyTo().
-            }
-            else
-                throw new NotImplementedException($"Not implemented type {typeof(T)}");
-            GC.KeepAlive(tensor);
+            var map = OnnxToMlNetTypeMap();
+            if (!map.ContainsKey(type))
+               throw Contracts.ExceptNotSupp("Onnx type not supported", type);
+            return PrimitiveType.FromKind(map[type]);
         }
 
-        public static PrimitiveType OnnxToMlNetType(DataType type)
+        internal static Dictionary<System.Type, DataKind> OnnxToMlNetTypeMap()
         {
-            DataKind kind;
-            switch (type)
+            if (_typeToKindMap == null)
```

Comment:
>_typeToKindMap [](start = 16, length = 14)

you can turn _typeToKindMap into static field/property so you don't need to check initialization.

Reply:
initialized it statically.

Reply:
Updated with static initilizer